walredo: fix zombie processes ([postgres] <defunct>)

problame opened the pull request:
This change wraps the std::process:Child that we spawn for WAL redo into a type that ensures that we try to SIGKILL + waitpid() on it.

If there is no explicit call to kill_and_wait(), the Drop implementation will spawns a task that does it in the BACKGROUND_RUNTIME. That's an ugly hack but I think it's better than doing kill+wait synchronously from Drop, since I think the general assumption in the Rust ecosystem is that Drop doesn't block.
Especially since the drop sites can be _any_ place that drops the last Arc<PostgresRedoManager>, e.g., compaction or GC.

The benefit of having the new type over just adding a Drop impl to PostgresRedoProcess is that we can construct it earlier than the full PostgresRedoProcess in PostgresRedoProcess::launch(). That allows us to correctly kill+wait the child if there is an error in PostgresRedoProcess::launch() after spawning it.

I also took a stab at a regression test. I manually verified that it fails before the fix to walredo.rs.

fixes https://github.com/neondatabase/neon/issues/2761

hlinnaka commented on test_runner/regress/test_walredo_no_zombies.py:
```diff
@@ -0,0 +1,102 @@
+        if not defunct_present:
+            assert child.status() != psutil.STATUS_ZOMBIE
+
+def test_walredo_no_zombies(neon_env_builder: NeonEnvBuilder):
```
```suggestion
# Check that the pageserver doesn't leave behind WAL redo processes
# when a tenant is detached. We had an issue previously where we failed
# to wait and consume the exit code of the WAL redo process, leaving it behind
# as a zombie process.
def test_walredo_not_left_behind_on_detach(neon_env_builder: NeonEnvBuilder):
```

This actually checks that there are no WAL redo processes, zombies or alive, left behind. Which makes perfect sense, but let's reflect that in the test name.

hlinnaka commented on pageserver/src/walredo.rs:
```diff
@@ -880,14 +852,95 @@ impl PostgresRedoProcess {
-    fn drop(&mut self) {
-        if !self.called_kill {
-            error!(tenant_id=%self.tenant_id, pid = %self.child.id(), "dropping PostgresRedoProcess that wasn't killed, likely causing defunct postgres process");
+/// Wrapper around st::process::Child that provides a
```
incomplete sentence

hlinnaka commented on pageserver/src/walredo.rs:
```diff
@@ -880,14 +852,95 @@ impl PostgresRedoProcess {
+    child: Option<Child>,
+}
+
+/// Wrapper type around `std::process::Child` which guarantees that the child
+/// will be killed and waited-for by this proces before being dropped.
+///
```
```suggestion
/// Wrapper type around `std::process::Child` which guarantees that the child
/// will be killed and waited-for by this process before being dropped.
```

hlinnaka commented on pageserver/src/walredo.rs:
```diff
@@ -880,14 +852,95 @@ impl PostgresRedoProcess {
+        Ok(NoLeakChild { child: Some(child) })
+    }
+
+    fn get_mut_child(&mut self) -> &mut Child {
+        self.child.as_mut().expect("must not call from drop")
+    }
+
```
How about implementing `Deref` and `DerefMut` for `Child` instead? I think that would eliminate the wrapper for `id` too.

hlinnaka reviewed (commented):
Are the changes in `poetry.lock` intentional? Necessary?

problame commented:
Addressed all review comments.
Re-did the poetry changes by using the upstream one (1.2.2) instead of the Fedora one (1.1.13).
That seems to have reduced the diff.
I don't understand why it's removing some of the  pyasn1 hashes from the lockfile.
My best guess is that the lockfile was generated by another version of poetry, and they changed their dependency resolution algorithm.

problame commented:
Squashed & address Python codestyle checks.

bayandin commented on test_runner/regress/test_walredo_not_left_behind_on_detach.py:
```diff
@@ -0,0 +1,106 @@
+    env = neon_env_builder.init_start()
+    pageserver_http = env.pageserver.http_client()
+
+    pagserver_pid = int((Path(env.repo_dir) / "pageserver.pid").read_text())
```
`env.repo_dir` is already a Path

```suggestion
    pagserver_pid = int(env.repo_dir / "pageserver.pid").read_text())
```

And then we don't need `from pathlib import Path` on the top of the file.

bayandin commented on test_runner/regress/test_walredo_not_left_behind_on_detach.py:
```diff
@@ -0,0 +1,106 @@
+        pageserver_http.tenant_detach(tenant_id)
+
+    # create new nenant
+    tenant_id, timeline_id = env.neon_cli.create_tenant()
```
We don't use this timeline_id right?

```suggestion
    tenant_id, _timeline_id = env.neon_cli.create_tenant()
```

bayandin commented on test_runner/regress/test_walredo_not_left_behind_on_detach.py:
```diff
@@ -0,0 +1,106 @@
+
+def assert_child_processes(pageserver_pid, wal_redo_present=False, defunct_present=False):
+    # XXX --ppid is probably GNU-only
+    children = list(psutil.Process(pageserver_pid).children())
```
I think it returns a list of processes by default

```suggestion
    children = psutil.Process(pageserver_pid).children()
```

LizardWizzard commented:
> My best guess is that the lockfile was generated by another version of poetry, and they changed their dependency resolution algorithm.

AFAIK it works like `cargo update` when you have incomplete versions specified. E g when you specify 1 as a version it can pick latest 1.X.X. So when you regenerate the lockfile poetry picks latest available version that satisfies requirements

problame commented:
Taking @knizhnik 's comments on this change from https://github.com/neondatabase/neon/pull/2778#issuecomment-1308776953 to here:

> But I do not see any temporary reason why child's stdin/sdout can not be taken and switched to non-blocking mode.
If it can happen, then it means some permanent OS/process problem which can not be somehow ignored.
If we try to continue normal work and spawn walredo process once agai then most likely we will be faced with the same error.
So I do not think that we should some how try to catch and handle this errors.

Yes, that's why I had `set_nonblocking(...).unwrap()` in an eariler PR, to make the pageserver die in those cases.
But @koivunej pointed out that
1. we don't like panicking the pageserver ( I think it's still the right thing to do in this case, but :shrug: )
2. since the panic happens in a thread other than the main thread, and main thread simply waits for a signal, it's not guaranteed that the pageserver is going to die if we panic here.

So, if we can't make the pageserver die from here, then we must somehow get rid of the child process that we already spawned.
The most elegant way to do that, in my opinion, is to re-use the `NoLeakChild` type added in this PR.
As explained in the commit message:

> The benefit of having the new type over just adding a Drop impl to PostgresRedoProcess is that we can construct it earlier than the full PostgresRedoProcess in PostgresRedoProcess::launch(). That allows us to correctly kill+wait the child if there is an error in PostgresRedoProcess::launch() after spawning it.

@knizhnik do you have objections to merging this?

knizhnik commented:
> @knizhnik do you have objections to merging this?

I do not want to block this PR but I still think that it is not needed.
There are 886 unwraps currently in Neon code. I am not sure that we need to eliminate them all just because panicking is not good thing. There are cases which should not happen and if them happen it is better not just ignore them but really stop work (may be by killing whole process).

I think that simplicity of code is more important than handling cases which never can happen;)

Concerning panicking in thread: it is really a problem. But it is not only related to this particular issue. There are many other cases when errors (including critical ones) are silently ignored in GC/compaction thread. Them just left messages in the log which nobody noticed. My suggestion was to treat all panics in background task i debug mode as fatal can kill pageserver in this case.
But @hlinnaka  prefer another solution: specify list of valid error messages for each test. But I do not thin that the problem is only related to regression tests failures. I still not convinced that ignoring (only logging) such errors in staging (and may be even in production) is the best thing we can do.

Sorry, it seems to be much bigger question than one addressed by this PR. If it will be enough just to set some flag then I will vote for this. But adding non so trivial peace of code sees to be overkill. But I do not insists. If people think that we should handle it, let's merge this PR and I will rebase my PR.

problame commented:
> There are cases which should not happen and if them happen it is better not just ignore them but really stop work (may be by killing whole process).

I wholeheartedly agree, whenever we are convinced that something cannot happen at runtime, we should panic the pageserver if it does.
But seems like the architecture doesn't permit that because panicking in the `::launch` method doesn't make the pageserver process exit.

> I think that simplicity of code is more important than handling cases which never can happen;)

Generally yes, but not when it comes to leaking global resources, because that eventually leads to failures in other components & other tenants.
In this case, zombie child processes are the resource,
Even more so if Rust uses pidfd some day, because then we'd be leaking file descriptors.
